crates/core/src/error.rs

```diff
@@ -375,6 +370,54 @@ impl RusticError {
     ) -> Box<Self> {
         Self::with_source(kind, error.to_string(), error)
     }
+
+    /// Returns a String representation for logging purposes.
+    ///
+    /// This is a more concise version of the error message.
+    pub fn display_log(&self) -> String {
+        let guidance = if self.context.is_empty() {
+            self.guidance.to_string()
+        } else {
+            // If there is context, we want to iterate over it
+            // use the key to replace the placeholder in the guidance.
+            let mut guidance = self.guidance.to_string();
+
+            self.context
+                .iter()
+                // remove context which has been used in the guidance
+                .for_each(|(key, value)| {
+                    let pattern = "{".to_owned() + key + "}";
+                    guidance = guidance.replace(&pattern, value);
+                });
+
+            guidance
+        };
+
+        let kind = format!("related to {}", self.kind);
+
+        let error_code = self
+            .error_code
+            .as_ref()
+            .map_or(String::new(), |code| format!(", code: {code}"));
+
+        let output = self.source.as_ref().map_or_else(
+            || {
+                format!("Error: {guidance} (kind: {kind}{error_code})")
+            },
+            |cause| {
+                cause.source().map_or_else(
+                    || {
+                        format!("Error: {guidance} (kind: {kind}{error_code}): caused by: {cause}")
+                    },
+                    |source| {
+                        format!("Error: {guidance} (kind: {kind}{error_code}): caused by: {cause} : (source: {source})")
+                    },
+                )
+            },
+        );
+
+        output
```

Comment:
Have you considered using a `fmt: std::fmt::Formatter` in an analog way as `impl Debug` or `impl Display` ?

I suggest that because your 3 branch here are formatted the same, it's just that some are appended.
A formatter could (maybe) simplify the code. (avoiding `map_or_else` and `closures` in favor of `if <cond> { fmt.write!("...")`

Reply:
Refactored to `write!(fmt , ..)`